[1단계 - 로또(자동)] 쿤(홍동건) 미션 제출합니다.

Hongdonggeon opened the pull request:
안녕하세요 스티치 🙌   쿤입니다!!
코드리뷰를 받을 수 있게 되어 너무 영광입니다 🤘🏻

열음과 페어 프로그래밍을 하면서 의문이 생겼던 부분에 대해서 몇가지 질문을 드리고 싶습니다!!

### 생성자 안에서 메서드 사용하기
아래 코드는 리팩토링 하기전에 Player 클래스입니다. 
`purchaseLotto()`에서 List<Lotto>를 생성하여 생성자안에서 멤버 필드에 주입해주는 방식이었습니다. 
생성자 외부에서 값을 받아 객체를 생성하지 않고, 해당 클래스 내부에서 값을 주입 받는 방식이 마음에 걸려서 리팩토링을 해보았는데요!
리팩토링 후에는,  purchaseLotto 메서드를 void를 반환형으로 바꿔주고, 맴버 필드 `List<Lotto> lottos`에 로또 객체 하나하나를 add해주는 식으로 리팩토링하였는데요!  어떤 방법이 더 괜찮은 방법일까요,,?
외부에서 값을 주입받아 객체를 만들어주지 않은 이유로는 로또 구매 시점보다 Player 객체 생성 시점이 더 빠르다고 생각했기 때문이예요!! 
이런 상황뿐만 아니더라도 생성자 안에서 메서드를 사용하여 값을 주입 받는 방법이 괜찮을지 궁금하네요!!


```java
public class Player {

    private Money money;
    private List<Lotto> lottos;

    public Player(Money money){
        this.money = money;
        this.lottos = purchaseLotto();
    }

    private int getNumberOfPurchases(){
        return money.determineQuantity();
    }

    public List<Lotto> purchaseLotto() {
        int numberOfPurchases = getNumberOfPurchases();
        List<Lotto> lottos = new ArrayList<>();
        for(int i = 0; i < numberOfPurchases; i++){
            lottos.add(new Lotto(LottoFactory.generateAutoLotto()));
        }
        return lottos;
    }
```



다른 질문들은 코드에 conversation으로 남겨놓겠습니다!!
감사드립니다 스티치!! 😀


https://brazen-serpent-b4f.notion.site/LV2-5410b558c47f47968f9610c7ffaa3bbf

Juhyung990122 committed:
docs: 기능 구현 목록 작성

Juhyung990122 committed:
feat: 입력 기능 구현

Juhyung990122 committed:
feat: Money 객체 생성

Juhyung990122 committed:
docs: 기능 구현 목록 수정

Juhyung990122 committed:
feat: LottoNumber 객체 생성

Juhyung990122 committed:
feat: 로또 구매 기능 구현

생성자 안에 복잡한 메소드가 들어가도 될지 고민되었습니다.

Juhyung990122 committed:
docs: 기능 구현 목록 수정

Juhyung990122 committed:
feat: DTO 추가

Juhyung990122 committed:
docs: 기능 구현 목록 수정

Juhyung990122 committed:
feat: WinningLotto 객체 생성

Juhyung990122 committed:
docs: 기능 구현 목록 수정

Juhyung990122 committed:
feat: Winning Lotto 당첨번호와 보너스볼 간 중복검사

Juhyung990122 committed:
refactor: 예외 메세지 별도 클래스로 분리

Juhyung990122 committed:
feat: 로또 구매 및 당첨번호 세팅 기능 구현

Juhyung990122 committed:
feat: 로또 판정 로직 및 전체 판정 구현

Juhyung990122 committed:
feat: 보너스볼 판정 로직 및 Result 객체 구현

Juhyung990122 committed:
feat: 2,3등은 보너스볼과 일치하는 숫자의 갯수를 기준으로, 나머지 등수는 일치하는 숫자의 갯수만으로 등수를 판정하는 로직 구현

Juhyung990122 committed:
feat: 결과 출력 기능 구현

Juhyung990122 committed:
feat: 수익률 계산 로직 및 출력 기능 구현

Juhyung990122 committed:
refactor: 수익률 계산 로직 분리 및 위치 변경

Juhyung990122 committed:
refactor: 프린트 할 결과물 생성 로직 변경

Juhyung990122 committed:
refactor: 결과 출력 로직 분리

Juhyung990122 committed:
refactor: 결과 리스트 생성 로직 이동

Juhyung990122 committed:
refactor: 매직넘버 상수화

Juhyung990122 committed:
refactor: 입력 유효성 검사 static으로 변경 및 적용

Juhyung990122 committed:
style: 코드 포멧팅

Juhyung990122 committed:
style: 코드포멧팅

Juhyung990122 committed:
refactor: 로또 생성 기능 인터페이스로 분리

Juhyung990122 committed:
style: 코드 포멧팅

Juhyung990122 committed:
refactor: 출력 개행 추가

Juhyung990122 committed:
refactor: 패키지 분리

Hongdonggeon commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.Scanner;
+import java.util.stream.Collectors;
+
+public class InputView {
```
### View 클래스들을 어떻게 사용하는게 좋은가?

저는 상태가 없는 클래스들은 모든 메서드를 static 으로 선언하여 유틸리티 클래스로 만들어주는데요!! static 메서드들을 너무 많이 사용하면 메모리 측면에서 안 좋아질까봐,, 약간 걱정이 되네요,, View를 인스턴스로 관리하는게 좋을지 지금처럼 유틸리티 클래스로 설계하여 사용하는게 좋을지 궁금합니다!!

lxxjn0 replied:
상태를 가지지 않는 클래스가 static으로 사용해도 되는 클래스인가에 대해서는 맞을 수도 있고 아닐 수도 있어요!! 
(이건 나중에 조금 더 학습하다보면 마주할 부분일 것 같아서 따로 코멘트를 남기진 않겠습니다, 예시로는 controller 클래스의 경우도 무상태 객체라고 볼 수 있어요!! 이는 MVC 패턴을 학습하다 보면 접할 수 있을 부분으로 보입니다 ㅎㅅㅎ)

저는 상태를 기준으로 판단하기 보다는 해당 클래스의 역할과 책임을 기준으로 static 매서드 사용 여부를 결정합니다!! 만약 해당 클래스가 유틸리티 성격이 강하고 객체로 생성해서 관리하기에는 사용되는 포인트가 다양할 경우엔 static 메서드로 활용하는 것 같아요!!

그러나 유틸리티 성격은 아닌데 인스턴스로 여러번 생성되기엔 성격이 적절하지 않다면 싱글톤 패턴을 사용하여서 인스턴스 생성을 제한할 수도 있을 것 같아요!! 해당 부분은 쿤이 View 클래스가 어떠한 성격에 가까운지 고민해보고 static 또는 싱글톤 패턴 등을 사용하여서 개선해보면 좋을 것 같습니다!!!
(저는 View가 유틸리티 성격인 것 같으면서도 아닌 것 같아서 어떠한 방법을 사용해도 괜찮을 것 같다는 생각은 듭니다!! ㅎㅅㅎ)

추가적으로 https://tecoble.techcourse.co.kr/post/2020-07-16-static-method/ 해당 글을 읽어보는 것도 도움이 될 것 같네요 👍🏼

Hongdonggeon replied:
View를 싱글턴 패턴으로 구현을 해봤습니다!! 모든 메서드가 static으로 관리되면 GC의 대상에서 벗어나기 때문에 메모리 효율에서 떨어질 수도 있기도 하고 고정된 메모리 영역을 얻으면서, 메모리 효율도 좋아지고 프로그램이 시작되면서 이미 인스턴스가 생성이 되서 속도가 빨라지는 장점이 있더라구요!! 

궁금한점으로 싱글턴 패턴을 사용할 때, 객체 생성할 때 static으로 선언을 하는데, 인스턴스가 static에서 관리가 된다면 해당 메서드들은 인스턴스가 static이더라도 GC의 대상이 되나요??

Hongdonggeon commented on src/main/java/dto/LottosDto.java:
```diff
@@ -0,0 +1,34 @@
+public class LottosDto {
+
+    int quantity;
+    List<LottoDto> lottoDtos;
```
Dto 안에 Dto로 감싸도 될까요? 모든 로또의 정보를 한번에 보내기 복잡할 것 같아서, Lotto 하나의 정보는 LottoDto, 사용자가 구매한 모든 로또의 정보는 LottosDto로 List<LottoDto>를 멤버 변수로 가지고 있는데요!! 이렇게 Dto를 두번 감싸도 문제가 없을지 궁금합니다!

lxxjn0 replied:
dto의 경우 반환하고자 하는 형식에 맞춰 생성할 수 있기 때문에 지금과 같이 한번 더 감싸는 형태도 상관없습니다!!

Hongdonggeon replied:
답변 감사합니다!! 현재로서는 저희가 Dto 를 보내고 꺼내는 작업을 하지만 추후에 프론트엔드와 협업을 할 때, Dto를 두번 감싸서 보내주어도 될까 해서 우려했었던 것 같아요,, 찾아보니까 자바 스크립트에서는 json 을 바로 객체로 인식해서 되는 것 같더라구요!!

Hongdonggeon commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,74 @@
+        }
+        return totalIncome;
+    }
+
```
Enum에 비즈니스 로직을 구현해도 괜찮을까요?? 총 수익을 계산하고, 등수 일치 횟수를 세는 로직을 Rank라는 Enum에서 구현한 이유로는, 비즈니스 로직에 필요한 값들이 Rank에 다 존재하기 때문에 효율적이라고 생각했고, 다른 객체에서 Rank를 참조하게 되면, 결합도가 높아지지 않을까 싶었습니다!

lxxjn0 replied:
해당 코드의 가장 큰 문제는 enum의 필드를 불변이 아닌 상태로 사용하고 있다는 점입니다. Enum은 상수로 관리되어야 되는 값이기 때문에 맴버 변수도 모두 불변으로 관리하는 것이 좋습니다!!

만약 로또를 2번 반복해서 구매하는 케이스가 등장한다면 현재 Rank는 한번 이상 사용할 수 없을 겁니다. Enum은 변경되지 않을 상수와 같이 관리되어야 하므로 만약 hitCount와 같이 가변으로 사용되어야 하는 부분, 비즈니스 로직은 외부로 뽑아서 관리하는 것이 적절할 것으로 보여져요!!

Enum은 상수 그 자체로, 나머지 비즈니스 로직은 적절한 도메인을 구현하거나 이용하여서 처리하는 것이 좋아 보입니다!!

Hongdonggeon replied:
Enum은 상수로 관리되어야 하기 때문에, 모든 멤버 변수도 불변으로 관리해야 하군요! 
구매한 로또의 당첨 통계를 관리하는 ResultStatus라는 클래스를 추가하였습니다!! Enum을 어떻게 사용해야할지 감을 잡은 것 같아요 친절한 답변 너무 감사드립니다 😀

Hongdonggeon committed:
fix: controller import 문 추가

Hongdonggeon committed:
refactor : 매직넘버 상수화

lxxjn0 commented on README.md:
```diff
@@ -1,7 +1,66 @@
-# java-lotto
+## 기능 구현 목록
```
꼼꼼하게 기능 구현 목록을 작성한 부분 좋습니다 👍🏼

lxxjn0 commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,21 @@
+
+        List<Result> results = lottoController.judgeLottos();
+        OutputView.printResult(lottoController.makeResult(results));
+    }
```
컨트롤러를 하나의 메서드만 제공하는 클래스가 아니라 적절한 단위로 메서드를 나눠서 사용한 부분 좋습니다!! 공백도 적절히 사용해서 코드의 흐름을 이해하기도 좋았어요 👍🏼 👍🏼

lxxjn0 commented on src/main/java/domain/Lotto/Lotto.java:
```diff
@@ -0,0 +1,42 @@
+
+    public List<LottoNumber> getLotto() {
+        return lotto;
+    }
```
외부에서 lotto 맴버 변수에 대한 조작을 최대한 제한하도록 반환해보는 것은 어떨까요?? Collections 클래스의 메서드를 사용하면 가능할 것 같습니다 :)

Hongdonggeon replied:
Collections.unmodifiableList()를 이용해 List안에 값들이 새로 추가되지 않도록 막아주었고, 일급 컬렉션 생성자에서 멤버 변수가 초기화 될 때,  new ArrayList()로 새로운 주소값을 할당받으면서 파라미터로 들어오는 List로 인해 값이 바뀌지 않게 막아주었습니다!

lxxjn0 commented on src/main/java/domain/Lotto/Lotto.java:
```diff
@@ -0,0 +1,42 @@
+        int hitCount = 0;
+        for (LottoNumber lottoNumber : lotto) {
+            hitCount += lottoNumber.checkHit(winningNumbers);
+        }
```
해당 로직을 Stream을 사용하여 개선한다면 보다 안전하게 불변으로 hitCount를 생성할 수 있을 것 같아요!!

Hongdonggeon replied:
넵 !! Stream의 filter()로 WinningLotto가 포함하고 있는 LottoNumber들로만 걸러주어서 count()라는 최종연산을 사용해서 갯수를 더 손쉽게 셀 수 있게 되었네요!! 가독성도 더 좋아진 것 같아요!! 정말 좋은 생각인 것 같아요 😀

lxxjn0 commented on src/main/java/domain/Lotto/LottoNumber.java:
```diff
@@ -0,0 +1,50 @@
+            return 1;
+        }
+        return 0;
+    }
```
해당 메서드 명과 반환 타입을 보면 어떠한 역할을 하는 메서드인지 파악하기 어려운 것 같아요!! 해당 메서드를 사용하는 곳을 확인해보면 이러한 방식이 아닌, WinningLotto에 메시지를 전달하는 방식으로 해결이 가능할 것 같은데 한 번 고민해보면 좋을 것 같아요!! :)

Hongdonggeon replied:
정말 좋은 방법인 것 같아요~~ !! WinningLotto에게  LottoNumber를 전달해 WinningLotto의 당첨번호가  포함하고 있는지와 List<LottoNumber>를 전달해 bonusBall을 포함하고 있는지 묻는 형식으로 메서드를 만들어주었어요!! 
값을 꺼내서 비교하는 형식이 아닌 메시지를 전달하는 방식으로 바꿔주면서 getter() 메서드도 없애줄 수 있게 되었네요!!

lxxjn0 commented on src/main/java/domain/Lotto/LottoNumber.java:
```diff
@@ -0,0 +1,50 @@
+
+    public int number() {
+        return this.number;
+    }
```
getter 양식을 따라보면 어떨까요?

Hongdonggeon replied:
반영했습니다!!

lxxjn0 commented on src/main/java/domain/Lotto/LottoNumber.java:
```diff
@@ -0,0 +1,50 @@
+    @Override
+    public int hashCode() {
+        return Objects.hash(number);
+    }
```
equals와 hashCode를 선언하지 말고 한 번 LottoNumber를 캐싱하는 전략을 가져가보면 어떨까요?? IntegerCache라는 클래스를 한번 참고해보면 도움이 될 것 같아요!! :)

Hongdonggeon replied:
로또 범위 안에서는 클래스 필드인 List<LottoNumber>에 있는 동일한 LottoNumber 인스턴스들을 모든 클래스에서 사용하게 해주면서, equals와 hashcode를 사용하지 않게 해줄 수 있게 되었네요!! 그리고 LottoNumber 객체를 여러번 생성하지 않게 되면서 메모리상으로도 되게 효율적인 방법이네요! 꿀팁 감사합니다 ☺️

lxxjn0 commented on src/main/java/domain/Lotto/LottoNumberFactory.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class LottoNumberFactory {
```
해당 부분도 만약 LottoNumber를 캐싱한다면 사라질 수도 있을 부분일 것 같아요!! 한 번 같이 고민해보면 좋을 것 같습니다!! 👍🏼

Hongdonggeon replied:
그렇네요!! LottoNumber의 클래스 필드인 List<LottoNumber> 를 모든 클래스에서 사용하게 해주면서, 따로 로또범위의 숫자의 List를 만들어주는 작업이 필요 없어졌네요!!

lxxjn0 commented on src/main/java/domain/Lotto/WinningLotto.java:
```diff
@@ -0,0 +1,33 @@
+        bonusBall = bonusBallNumber;
+    }
+
+    public void validateDuplicate(List<LottoNumber> numbers, LottoNumber bonusBallNumber) {
+        if (numbers.contains(bonusBallNumber)) {
+            throw new IllegalArgumentException(ExceptionMessage.LOTTO_AND_BONUS_BALL_DUPLICATION);
+        }
+    }
```
꼼꼼한 유효성 검증 굿 👍🏼

lxxjn0 commented on src/main/java/domain/LottoGenerator/AutoLottoGenerator.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class AutoLottoGenerator implements LottoGenerator {
```
로또 번호 생성기를 따로 구현하여서 로또 생성의 책임을 지게 하였군요!! 추후 확장성도 고려한 코드로 보여지네요!! 💯

lxxjn0 commented on src/main/java/domain/LottoGenerator/AutoLottoGenerator.java:
```diff
@@ -0,0 +1,27 @@
+                .collect(Collectors.toList());
+
+        return new Lotto(lottoNumbers);
+    }
```
현재는 사용할 수 있는 숫자를 넘겨주고 로또를 생성하고 있습니다. 하지만 해당 메서드 명만 본다면 '숫자들을 받아서 로또를 만든다' 정도로 읽을 수 있어요!! 
그렇다면 해당 메서드를 사용하는 사용자는 어떻게 사용해야지 올바르게 사용할 수 있는지(45개의 숫자를 인자로 넘겨줘야 한다)를 알 수 있을까요?? 해당 메서드를 조금 더 개선해보면 좋을 것 같아요!! (앞서 캐싱 전략을 적용한다면 해당 부분도 좀 더 적절하게 개선이 가능할 것 같네요!!)

lxxjn0 commented on src/main/java/domain/LottoGenerator/WinningLottoGenerator.java:
```diff
@@ -0,0 +1,20 @@
+
+        return new Lotto(lottoNumbers);
+    }
+}
```
해당 클래스는 Lotto의 정적 팩토리 메서드로도 해결이 가능해 보이는데 쿤의 생각은 어떤가요?

Hongdonggeon replied:
제 생각에는 당첨 번호도 로또와 똑같다고 생각을 했어요!! 현재로서는 로또 자동생성 기능만 있지만, 당첨번호를 뽑는 것도 로또의 수동 생성과 똑같은 방법이라고 생각을 했어요!! 그래서 LottoFactory에 당첨 번호를 생성해주는 새로운 메서드를 만드는 것보다 로또를 생성해주는 LottoGenerator에서 구현체를 만들어주는 것이 더 적합하다고 생각했어요!

Hongdonggeon replied:
지금 LottoFactory에서 입력받은 List<String>을 List<Integer>로 바꿔주는 메서드가 있는데, 다시 생각해보니 View에서 충분히 해줄 수 있다는 생각이 드는 것 같아요!! View에서 입력한 값을 Controller에게 보내줄 때, 로또 게임에 어울리는 형식에 맞춰 데이터를 최소한의 정리해주는 것은 충분히 View에서  할 수 있다는 생각이 들었던 것 같아요.

lxxjn0 replied:
위에서 쿤이 적어준 생각도 충분히 합리적인 것 같아요!! 해당 부분은 쿤의 생각에 맞게 코드를 작성해봐도 좋을 것 같아요!!

Hongdonggeon replied:
AutoLottoGenerator에서는 로또가 자동으로 생성이 되어야 하기에 parameter가 없게, ManualLottoGenerator에서는 사용자가 로또 번호를 고르는 방법이기에 List<Integer>라는 파라미터를 두었습니다. 
그리고 각 클래스에서 인터페이스에서 원하는 메서드만 오버라이딩을 하기 위해 LottoGeneratorAdpapter라는 클래스를 추가했습니다.

lxxjn0 commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,74 @@
+    FOURTH(4, 50000, 0),
+    THIRD(5, 1500000, 0, false),
+    SECOND(5, 30000000, 0, true),
+    FIRST(6, 2000000000, 0);
```
얻을 수 있는 등수를 enum으로 관리하였군요!! 👍🏼

lxxjn0 commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,74 @@
+        if (result.compare(value)) {
+            value.hitCount++;
+        }
+    }
```
위에서 말한 것과 같이 해당 로직의 경우 enum 맴버 변수를 변경하면서 로직을 구현하였어요!! 
Rank가 순전히 상금을 얻을 수 있는 조건을 관리하는 책임을 가진다면 지금과 같은 로직은 '상금을 얻을 수 있는 조건'이 아닌 '얼마의 상금을 얻을 수 있는지를 연산하는 책임'까지 가지고 있는 것으로 보이네요!!

물론 금액을 연산하는 로직을 가지는 것은 잘못된 것이 아닙니다. 내부에서 각각의 순위에 일치하는 개수 등이 가변 맴버 변수를 통해 관리되고 있는 부분이 개선이 필요할 것으로 보여집니다!! 아래의 글을 한번 학습해보는 것도 좋을 것 같아요!! 👍🏼 

https://techblog.woowahan.com/2527/

Hongdonggeon replied:
위에 글을 보고 Enum의 사용 목적, 어떻게 사용해아 할지에 대해 감을 잡을 수 있었고, 함수형 인터페이스를 필드에 추가하여 사용하는 것도 너무 신기하게 봤습니다!! 좋은 글 감사드립니다!!

lxxjn0 commented on src/main/java/domain/player/Player.java:
```diff
@@ -0,0 +1,47 @@
+        for (int i = 0; i < numberOfPurchases; i++) {
+            lottos.add(lottoGenerator.generateLotto(lottoNumbers));
+        }
+    }
```
질문에 대한 답변을 여기에 남기도록 하겠습니다!! 먼저, 저의 생각은 현재와 같은 코드가 더 나은 방향이라고 생각돼요!! 그러나 조금만 더 깊게 생각해보면 좋을 것 같아요.

먼저 Player가 가지고 있는 맴버 변수를 통해 Player의 책임을 생각해보자면 '처음에 입력한 Money를 가지면서, 이를 사용해 구매한 Lotto를 관리하는 책임' 정도라고 생각됩니다. 

그런 관점에서 다시 해당 메서드를 살펴볼까요? 해당 메서드를 보면 '로또 생성기랑 로또 숫자들을 줄테니 너가 로또를 구매해' 의 행위를 나타낸다고 보여져요. 본인이 처음 입력한 돈을 관리하고, 이를 사용해 구매한 로또를 관리하는 책임을 기준으로 본다면 해당 메서드의 행위는 정말 Player가 가져야 하는 행위인가? 하는 생각이 들긴 합니다. 

차라리 로또 생성기에게 '내가 가지고 있는 money를 줄게, 로또를 생성해서 준다면 내 로또 보관함에 추가할거야!' 와 같이 행위를 하는게 좀 더 적절하지 않을까? 하는 생각도 들어요. 그리고 Money가 controller와 같은 메서드에서 관리가 된다면 Money 금액 자체가 그대로 유지되어도 딱히 이상은 없을 것 같습니다!! 단지 구매 가능한 금액을 저장한 정도로 볼 수 있으니깐요. 

그러나 Player가 Money를 가지게 된다면 로또를 구매하면서 가지고 있는 Money의 금액도 줄어야 하지 않을까요?? 😄 

이런 여러 고민들을 해보면서 해당 로직을 다시 한번 개선해보면 좋을 것 같아요!! 객체는 자신이 가지는 맴버 변수가 곧 자신의 존재 이유이자 책임이라고 볼 수 있습니다! 그런 의미에서 Player가 money랑 lottos를 가지게 된다면, Player는 어떠한 책임을 가지는 객체가 되어야 하는지도 고민해보면 좋을 것 같아요!!

Hongdonggeon replied:
스티치의 말에 공감이 됩니다. 리팩토링 하기전에, Player의 책임에 대해 살펴봤을 때, 로또 생성기와 숫자를 받아서 Player가 로또를 생성해서 로또 보관함에 추가하는 역할을 하더라구요! 설계하면서 생각했던 Player 역할과는 조금 다른 것 같았어요! 

그래서 Money에 돈을 차감하는 메서드와 Player에 로또를 살수있는지의 여부와, 로또를 구매하는 기능을 추가시켜주었습니다.
로또를 생성하는 역할은 LottoGenerator와 Controller에서 하는게 더 알맞겠다는 생각이 들었습니다.

하지만, 고민이 되는 부분이 purchaseLotto라는 메서드가 너무 settter 메서드의 색깔을 지니고 있다는 생각이 드네요,, 이런 고민을 하면서 또한 Player 객체가 정말 필요할까라는 생각도 들었던 것 같아요. 로또 게임에 있어서 Player 객체가 여러번 생성될 일이 없기 때문에, 굳이 필요 없겠다는 생각이 드네요. 차라리 Money는 정말 Controller 메서드에서 관리가 되고, 차라리 `List<Lotto>`를 관리하는 Lottos라는 일급 컬렉션을 만들까도 고민이 정말 되었던 것 같아요. 
하지만, 제가 처음 설계한 생각대로 리팩토링 해보고 싶었기에 해보았습니다.

스티치는 기능 요구 사항을 보고, 어떤 도메인이 있을까 고민하는 과정속에서 어떤 기준으로 도메인을 나누시는지 정말 궁금합니다!!

lxxjn0 commented on src/main/java/dto/LottoDto.java:
```diff
@@ -0,0 +1,26 @@
+
+    public LottoDto(List<Integer> lottoNumber) {
+        this.lottoNumber = lottoNumber;
+    }
```
외부에서 사용하지 않는다면 private으로 접근 지정자를 설정하는 것도 좋을 것 같아요!!! 그렇게 하면 어떠한 장점이 있을까요??

Hongdonggeon replied:
외부에서 Dto 객체 생성을 막아줄 수 있게 됩니다!! 유틸리티 클래스를 만들때도 기본 생성자를 private으로 변경 해주면 좋겠네요!!

lxxjn0 commented on src/main/java/dto/LottoDto.java:
```diff
@@ -0,0 +1,26 @@
+                .map(LottoNumber::number)
+                .collect(Collectors.toUnmodifiableList());
+        return new LottoDto(numbers);
+    }
```
해당 메서드도 같은 패키지 내에서만 사용되는 것으로 보이는데 그에 맞게 접근 지정자를 설정해주면 좋을 것 같아요!!

Hongdonggeon replied:
같은 패키지내에서만 사용되는 메서드이므로 default로 변경해주어서 다른 패키지에서는 메서드를 사용하지 못하게 막아주었습니다!!

lxxjn0 reviewed (changes requested):
안녕하세요, 쿤 :)
이번에 쿤의 리뷰를 맡게 된 스티치입니다!!

미션 구현 코드 잘 봤습니다!! 최대한 각각의 책임에 맞게 도메인을 분리하고 관리하려고 노력한 모습이 잘 느껴졌어요!!

이번 리뷰에서는 도메인 각각의 책임에 맞게 구현되어 있는지를 중점적으로 코멘트를 남겼습니다!! Enum의 사용 방법에 대해서도 조금 개선이 필요할 것 같아서 해당 부분에 집중해서 코멘트를 반영해보면 좋을 것 같아요 👍🏼 

남겨주신 질문은 해당 코드에 코멘트로 남겨두었습니다!! 리뷰 확인 부탁드릴게요 :)

Hongdonggeon committed:
refactor : 일급 컬렉션의 List 불변 객체로 만들기

생성자의 파라미터로 받는 List를 그대로 멤버 변수에 주입하지 않고, new ArrayList()로 새로운 주소값을 할당받음
Collections.unmodifiableList()를 이용해 밖에서 List의 값이 바뀌지 않게 막는다.

Hongdonggeon committed:
refactor : Dto 생성자 private 으로 접근지정자 변경

Dto 생성자를 자기 자신의 클래스에서만 사용하기 때문에 private 접근지정자로 변경해줌
외부에서 Dto 객체 생성해주는것도 막아줄 수 있다.

Hongdonggeon committed:
refactor : 메서드 접근지정자 default로 변경

LottoDto, RankDto의 from 메서드가 같은 패키지내에서만 사용하기 때문에 접근 지정자를 default로 변경해줌

Hongdonggeon committed:
refactor : 당첨 로또와 사용자 로또 비교하는 메서드와, 보너스볼과 사용자 로또 비교해주는 메서드 위치 변경

WinningLotto 객체를 가져와서 당첨번호와 보너스볼이란 값을 꺼내서 비교하는 것이 아닌, WinningLotto 객체에게 LottoNumber 를 포함하고 있는지 메시지를 보내주는 형식으로 변경

Hongdonggeon committed:
refactor : List<String>을 List<Integer> 로 변환해주는 역할 View에 위임

View의 역할은 구분자를 분리하고, 공백을 제거해주는 등과 같은 데이터를 최소한의 정리된 형태로 만들어, Controller로 넘겨주는 것까지라고 생각된다.

Hongdonggeon committed:
feat : 반복되는 LottoNumber 인스턴스 캐싱하기

LottoNumber 1에서 45까지 클래스 필드 List<LottoNumber>에 추가하여, LottoNumber 객체가 필요할 때, valueOf() , values()로 꺼내서 사용한다.
모든 클래스는 같은 LottoNumber 인스턴스를 사용하므로, equals와 hashcode도 필요없게 되었다

Hongdonggeon committed:
remove : LottoNumberFactory 삭제

LottoNumber의 List<LottoNumber>를 모든 클래스에서 사용하게 하면서, 따로 1 ~45 의 숫자를 리스트에 넣어주는 작업이 필요 없어졌다

Hongdonggeon committed:
refactor : Player 로또 구매 역할 분리

Player에 로또를 구매할 수 있는지여부와 로또를 구매하는 역할추가
로또 구매 기능에 있어서 Player는 전달받는 (생성된)로또를 로또 보관함에 추가하한다.

Controller에서 LottoGenerator를 이용하여 로또를 생성한다.

Hongdonggeon committed:
feat : 구매한 모든 로또의 결과를 기록하는 클래스 추가

이전에는 Enum에서 hitcount라는 변수를 사용하여, 로또의 Rank 기록들을 관리하였는데 Enum의 필드는 모두 상수로 관리되어야 한다.
그래서 ResultStatus라는 당첨 통계를 기록하고 관리하는 클래스를 추가하였다.

Hongdonggeon committed:
refactor : InputView, OutputView 싱글턴 패턴으로 관리

view의 모든 메서드들이 static 메서드로 선언이 되어있었는데, static은 GC의 대상이 아니기 때문에 view의 인스턴스는 static으로 관리하고 메서드들은 인스턴스로 관리한다.

Hongdonggeon committed:
refactor : 인터페이스에서 원하는 추상 메서드만 몸체를 구현하고자 어댑터 클래스 추가

인터페이스 추상 메서드를 오버로딩하여 선언하였고 인터페이스를 원하는 메서드만 구현하고자 어댑터 클래스를 추가하였다.
AutoLottoGenerator에서는 parameter가 없는 generateLotto만, ManualLottoGenerator에서는 parameter가 있는 generateLotto를 오버라이딩 하였따

lxxjn0 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,45 @@
+    private Player player;
+    private WinningLotto winningLotto;
+
+    public LottosDto purchase(int purchaseAmount) {
+        player = new Player(purchaseAmount);
+        LottoGenerator lottoGenerator = new AutoLottoGenerator();
+        while (player.canBuyLotto()) {
+            player.purchaseLotto(lottoGenerator.generateLotto());
+        }
+        return LottosDto.from(player.getLottos());
+    }
```
현재는 LottoGenerator로 AutoLottoGenerator를 purchase 함수 내부에서 생성하고 있어요. 그렇다면 만약 LottoGenerator를 Auto 대신 Manual로 변경한다면 해당 메서드를 고쳐야 하지 않을까요??
해당 메서드를 고치지 않고도 이후 LottoGenerator를 변경할 수 있는 방법으로 한번 개선해보면 좋을 것 같아요!!

lxxjn0 commented on src/main/java/domain/Lotto/Lotto.java:
```diff
@@ -0,0 +1,39 @@
+
+public class Lotto {
+
+    public static final int CORRECT_LOTTO_SIZE = 6;
```
해당 변수가 Lotto 내부에서만 사용하는 것 같은데 접근지정자를 수정하면 좋을 것 같아요!!

Hongdonggeon replied:
아 그렇네요! 리뷰 반영했습니다 ☺️

lxxjn0 commented on src/main/java/domain/Lotto/LottoNumber.java:
```diff
@@ -0,0 +1,50 @@
+
+public class LottoNumber {
+
+    private static final int MINIMUM_LOTTO_NUMBER = 1;
+    private static final int MAXIMUM_LOTTO_NUMBER = 45;
+    private static final List<LottoNumber> CACHE = new ArrayList<>();
+
+    static {
+        for (int i = MINIMUM_LOTTO_NUMBER; i <= MAXIMUM_LOTTO_NUMBER; i++) {
+            CACHE.add(new LottoNumber(i));
+        }
+    }
```
캐싱을 적용해서 인스턴스 재사용을 가능하게 하였네요!! 👍🏼

lxxjn0 commented on src/main/java/domain/Lotto/LottoNumber.java:
```diff
@@ -0,0 +1,50 @@
+        this.number = number;
+    }
+
+    public static LottoNumber valueOf(int number) {
+        LottoNumber lottoNumber = CACHE.get(number - 1);
+
+        if (lottoNumber == null) {
+            lottoNumber = new LottoNumber(number);
+        }
+        return lottoNumber;
+    }
```
현재는 index에 의존적인 조회가 이뤄지는 것 같아요!! 만약 CACHE에서 1에 해당하는 LottoNumber가 삭제됐다가 다시 등록된다면 현재의 메서드는 사용이 불가능할 것 같습니다!! 다른 자료구조를 사용해서 이러한 문제점에 자유로울 수 있도록 개선해보면 좋을 것 같아요!!

Hongdonggeon replied:
맞아요! 이 부분에 대해 고민이 있었어요,, List를 Map으로 바꿔주고 valueOf()에서 map에서 꺼낸값이 null 값일때, cache 역할을 하는 map에 저장이 되게했습니다! 어차피 1~45에서 벗어난 숫자를 valueOf()에서 사용할시에 LottoNumber객체를 생성하는 도중에 예외가 발생할테니까요!!

lxxjn0 commented on src/main/java/domain/Lotto/WinningLotto.java:
```diff
@@ -0,0 +1,40 @@
+        return lottoNumbers.contains(bonusBall);
+    }
+
+    public void validateDuplicate(List<LottoNumber> numbers, LottoNumber bonusBallNumber) {
+        if (numbers.contains(bonusBallNumber)) {
+            throw new IllegalArgumentException(ExceptionMessage.LOTTO_AND_BONUS_BALL_DUPLICATION);
+        }
+    }
```
코드가 사용하는 곳에 가까이 붙어있으면 가독성이 더 좋아집니다!! 현재는 생성자에서 사용되는 검증 메서드가 생성자와 멀리 떨어지면서 가독성이 조금 떨어지는 것 같아요!!!
코드를 읽을 때 자연스럽게 아래로 읽을 수 있도록 배치를 해보면 가독성도 좋고 코드를 보기에 더 빠를 것 같아요!!

Hongdonggeon replied:
뭔가 validate 메서드는 비즈니스 로직에는 벗어나는 메서드라고 생각해서 마지막에 배치해두는 습관이 있었네요!! 확실히 사용되는 곳 바로 아래 메서드를 배치하니까 가독성이 더 좋아진 것 같아요! 꿀팁 감사합니다~~

lxxjn0 commented on src/main/java/domain/Lotto/WinningLotto.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public boolean isContainLottoNumber(LottoNumber lottoNumber) {
+        List<LottoNumber> winningLottoNumbers = winningLotto.getLotto();
+        return winningLottoNumbers.contains(lottoNumber);
```
Lotto 클래스에 이러한 질의를 하는 메서드를 만들어보면 어떨까요?? 지금처럼 getter로 가져와서 로직을 실행하는 것보다 더 객체지향적인 구조가 될 것 같아요!!

Hongdonggeon replied:
Lotto에 메서드를 추가하여 Lotto객체에 메세지를 보내는 구조로 변경하였습니다!

lxxjn0 commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,51 @@
+    private final int reward;
+    private boolean hitBonusBall;
+
+    Rank(int rankNumber, int criteria, int reward) {
+        this.rankNumber = rankNumber;
+        this.criteria = criteria;
+        this.reward = reward;
+    }
+
+    Rank(int rankNumber, int criteria, int reward, boolean hitBonusBall) {
+        this(rankNumber, criteria, reward);
+        this.hitBonusBall = hitBonusBall;
+    }
```
현재 생성자 오버로딩의 구조가 조금 어색한 것 같아요!! 객체의 모든 맴버 변수를 초기화하는 생성자를 주 생성자로, 제한된 맴버 변수를 받는 생성자를 보조 생성자의 개념으로 작성해보면 좋을 것 같아요!!

Hongdonggeon replied:
isSecondRank라는 메서드를 추가해서 hitBonusBall 인스턴스 변수를 제거해줄 수 있었어요!!

lxxjn0 commented on src/main/java/domain/player/Player.java:
```diff
@@ -0,0 +1,49 @@
+
+public class Player {
+
+    private static final int MINIMUM_PURCHASE_AMOUNT = 1000;
```
private 상수로 Money, Player 양쪽에서 모두 관리되고 있는 것 같아요. 적절한 곳에서 1개의 변수로 관리하는 것이 좋을 것 같아요!!!

Hongdonggeon replied:
같은 패키지내에서만 사용하고 있어서 Money에서 접근지정자를 default로 변경해주었습니다

lxxjn0 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,62 @@
+    private static final String BONUS_BALL_INPUT_MESSAGE = "보너스 볼을 입력해 주세요.";
+    private static final String DELIMITER = ",";
+
+    private static final Scanner scanner = new Scanner(System.in);
+
+    private static final InputView inputView = new InputView();
```
변수명을 상수에 맞게 선언해주면 좋을 것 같아요!!

Hongdonggeon replied:
리뷰 반영했습니다!!

lxxjn0 commented on src/main/java/utils/InputValidator.java:
```diff
@@ -0,0 +1,20 @@
+package utils;
+
+public class InputValidator {
```
혹시 해당 클래스를 따로 생성한 이유가 있을까요?? 해당 메서드만 관리하기엔 역할이 적어보여서 InputView에서 메서드로 관리하면 조금 더 응집도있게 view를 관리할 수 있을 것 같은데 쿤의 생각은 어떠신가요??

Hongdonggeon replied:
그렇네요! 스티치 말처럼 지금 InputValidator의 역할은 InputView에서 하는게 더 좋을 것 같습니다

lxxjn0 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,61 @@
+
+public class OutputView {
+
+    public static final String PURCHASE_QUANTITY_FORMAT = "%s개를 구매했습니다.%n";
```
해당 변수도 접근 지정자를 수정해도 될 것 같아요!!

Hongdonggeon replied:
넵!!

lxxjn0 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,61 @@
+    private static final String EMPTY_FORMAT = "";
+    private static final String INCOME_RATE_FORMAT = "총 수익률은 %.2f입니다.%n";
+
+    private static final OutputView outputView = new OutputView();
```
해당 변수도 상수에 맞는 변수명을 지정해주시면 좋을 것 같습니다!!

Hongdonggeon replied:
넵~~!!

lxxjn0 commented on src/main/java/domain/LottoGenerator/LottoGeneratorAdpater.java:
```diff
@@ -0,0 +1,17 @@
+
+import java.util.List;
+
+public class LottoGeneratorAdpater implements LottoGenerator{
```
이번에 adapter라는 클래스를 구현해줬는데 해당 클래스가 하는 역할은 인터페이스에서도 충분히 가능할 것 같아요!! 인터페이스의 default 메서드에 대해 한번 찾아보시면 좋을 것 같아요!!
추가적으로 특정 Generator에서 사용하지 못하는 메서드의 경우 적절한 예외를 발생하도록 하는 것도 좋을 것 같습니다!!

Hongdonggeon replied:
기존의 로또 자동 생성기능은 추상 메서드로 구현하고 새로 추가된 로또 수동생성 메서드는 default 메서드로 구현하였어요!!
AutoLottoGenerator와 ManualLottoGenerator에서 알맞지 않은 기능을 사용할시 예외 발생하도록 추가하도록 오버라이딩 하였습니다!

lxxjn0 commented on src/main/java/domain/player/Player.java:
```diff
@@ -0,0 +1,49 @@
+        return money.isBiggerThanLottoPrice();
+    }
+
+    public void purchaseLotto(Lotto lotto) {
+        money.deductMoney();
+        lottos.add(lotto);
+    }
```
현재는 해당 메서드에서 유효성 검증이 없는 것 같아요!! 해당 메서드가 사용 가능한 상태인지 유효성을 검증하는 로직도 추가하면 좋을 것 같아요!!

Hongdonggeon replied:
아 그렇네요 while문에 조건을 검사해주고 있지만, while문 밖에서 purchaseLotto를 이용할시에 돈이 부족해도 구매가 되겠네요!! 오 여기까지 생각 못했던 것 같아요 감사합니다~~

lxxjn0 commented on src/main/java/domain/player/Money.java:
```diff
@@ -0,0 +1,33 @@
+
+    private int amount;
+
+    public Money(int amount) {
+        validateAmount(amount);
+        this.amount = amount;
+    }
```
현재는 구매 금액에 1000원 이상이어야 한다는 만족하지만 1000원 단위의 구매금액이어야 하는지는 검증이 안될 것 같아요!! 해당 검증도 추가하면 좋을 것 같습니다!!

Hongdonggeon replied:
처음에 기능 요구사항 작성할 때, 1000원 단위로 입력하지 않을 경우 예외를 발생해야 하나 페어랑 논의를 했는데 예를 들어 2300원을 입력하면 2000원은 구매되고 300원은 그대로 되돌려받을 수 있지 않을까라고 생각했습니다! 하지만 현재 거스름돈 기능이 없기 때문에 1000원 단위로 입력하지 않은 경우에 예외 발생하도록 하였습니다

lxxjn0 commented on src/main/java/domain/player/Player.java:
```diff
@@ -0,0 +1,49 @@
+        return result;
+    }
+
+    public double calculateIncomeRate(double totalIncome) {
+        return totalIncome / (lottos.size() * MINIMUM_PURCHASE_AMOUNT);
+    }
```
현재 money에서는 1000원 단위의 구매 금액을 받지 못하기 때문에 해당 메서드가 올바른 결과를 반환하지 못할 것 같아요!! Money에 유효성을 검증하는 로직을 추가로 두고 진행한다면 해당 로직도 참의 결과를 반환할 수 있을 것 같습니다!!

Hongdonggeon replied:
로또를 구매할 경우에 money가 감소되게 구현을 했기 때문에 모두 로또를 구매하면 money가 0이 되어서 분모에 money를 넣지 않고, `구매한 로또수 x 로또 가격` 으로 하였는데, money가 감소되는 기능을 구현하기 전에는 제대로 된 수익률을 반환하지 못했겠네요!

lxxjn0 commented on src/main/java/dto/RankDto.java:
```diff
@@ -0,0 +1,38 @@
+
+import domain.Rank;
+
+public class RankDto {
```
해당 클래스는 dto로 만들어야 하는 목적이 따로 존재하지 않아 보여요! 기존의 클래스 형식 그대로 반환하고 있는데 dto로 구현한 이유가 있을까요?

Hongdonggeon replied:
기존의 클래스 형식 그대로 반환을 하고 있지만, VIEW와 Controller에서 독립적인 구조를 만들어주기 위함이였어요!! view가 현재처럼 하나의 application에 포함되어 있는 것이 아닌 view와 Controller에서 api로 통신을 하는 경우에, 통계 결과 출력에 필요한 데이터들을 한번의 통신으로 모두 보내주기 위헤 RankDto라는 클래스를 추가하였습니다.

lxxjn0 reviewed (approved):
쿤, 리뷰 반영해주신 부분 잘 보았습니다!!
대부분 적절하게 리뷰 반영해주시고 개선된 코드를 작성해주셨네요 :)
이번 피드백을 참고해서 2단계 미션 이어서 진행해주시면 될 것 같습니다!! 1단계 진행하느라 수고 많으셨어요!! 👍🏼
